app/src/main/java/com/andrew/tasky/presentation/task_detail/TaskDetailFragment.kt

```diff
@@ -4,29 +4,198 @@ import androidx.lifecycle.ViewModelProvider
+import kotlinx.coroutines.flow.collectLatest
+import java.text.SimpleDateFormat
+import java.util.*
 
 class TaskDetailFragment : Fragment(R.layout.fragment_task_detail) {
 
+    private val calendar = Calendar.getInstance()
     private lateinit var viewModel: TaskDetailViewModel
     private lateinit var navController: NavController
     private lateinit var binding: FragmentTaskDetailBinding
-
-    override fun onActivityCreated(savedInstanceState: Bundle?) {
-        super.onActivityCreated(savedInstanceState)
-        viewModel = ViewModelProvider(this).get(TaskDetailViewModel::class.java)
-    }
+    private lateinit var taskType: TaskType
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
         binding = FragmentTaskDetailBinding.bind(view)
         navController = Navigation.findNavController(view)
+        viewModel = ViewModelProvider(this).get(TaskDetailViewModel::class.java)
+        taskType = TaskType.values()[arguments?.getInt("taskType")!!]
+
+        subscribeToObservables()
+        setupTaskType()
+
+        if(viewModel.getEditMode()){
+            editMode()
+        }
+        else{
+            nonEditMode()
+        }
+
+        binding.closeButton.setOnClickListener {
+            navController.popBackStack()
+        }
+
+        binding.editButton.setOnClickListener {
+            viewModel.setEditMode(true)
+            editMode()
+        }
+
+        binding.saveButton.setOnClickListener {
+            navController.popBackStack()
+        }
+
+        binding.taskTitle.setOnClickListener {
+            val editType = EditType.TITLE.ordinal
+            val bundle = bundleOf("editType" to editType)
+            navController.navigate(R.id.action_taskDetailFragment_to_editFragment, bundle)
+        }
+
+        binding.editTitleButton.setOnClickListener {
+            val editType = EditType.TITLE.ordinal
+            val bundle = bundleOf("editType" to editType)
+            navController.navigate(R.id.action_taskDetailFragment_to_editFragment, bundle)
+        }
+
+        binding.editDescriptionButton.setOnClickListener {
+            val editType= EditType.DESCRIPTION.ordinal
+            val bundle = bundleOf("editType" to editType)
+            navController.navigate(R.id.action_taskDetailFragment_to_editFragment, bundle)
+        }
+
+        binding.descriptionTextView.setOnClickListener {
+            val editType= EditType.DESCRIPTION.ordinal
+            val bundle = bundleOf("editType" to editType)
+            navController.navigate(R.id.action_taskDetailFragment_to_editFragment, bundle)
+        }
+
+        binding.time.setOnClickListener {
+            timePickerFragment()
+        }
+
+        binding.date.setOnClickListener {
+            datePickerFragment()
+        }
+    }
+
+    private fun subscribeToObservables() {
+        lifecycleScope.launchWhenStarted {
+            viewModel.selectedDate.collectLatest {
```

Comment:
1. In fragments, always use `viewLifecycleOwner.lifecycleScope` which is tied to the lifecycle of what's visible (you only want to update views when they are actually active)
2. It is recommended to use the `repeatOnLifecycle` API to collect Flows in the UI. Not the end of the world for StateFlows like here, but as soon as you deal with events (for example coming from a SharedFlow), these can get lost if you're unlucky.
The reason is that lifecycleScope is cancelled when `onStop()` is reached. Therefore, if an event comes in between `onStop()` and `onDestroy()`, it would be lost since there are no active observers.

Reply:
https://medium.com/androiddevelopers/a-safer-way-to-collect-flows-from-android-uis-23080b1f8bda